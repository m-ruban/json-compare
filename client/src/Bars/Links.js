import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import { Nav, NavItem, Collapse, NavLink, NavbarToggler } from 'reactstrap'
import DropdownTests from './Examples/DropdownTests'
import withToggleState from '../HOC/withToggleState'

const Links = ({isOpen, toggle, feedback}) => (
  <>
    <NavbarToggler onClick={toggle} />
    <Collapse isOpen={isOpen} navbar>
      <Nav className="ml-auto" navbar>
        <DropdownTests />
        <NavItem>
          <Link className="nav-link" to="/faq/">FAQ</Link>
        </NavItem>
        <NavItem>
          <NavLink
            className="nav-link"
            href={feedback}
            target="_blank"
          >
            Feedback
          </NavLink>
        </NavItem>
      </Nav>
    </Collapse>
  </>
)

Links.propTypes = {
  feedback: PropTypes.string
}

const LinksWithToggleState = withToggleState(Links)

export default LinksWithToggleState
