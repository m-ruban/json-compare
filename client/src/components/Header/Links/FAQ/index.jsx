import React, { useState, useCallback, Fragment } from 'react';
import {
    Button,
    Modal,
    ModalHeader,
    ModalBody,
    ModalFooter,
    ListGroupItem,
    ListGroupItemHeading,
    ListGroupItemText,
    Badge,
    ListGroup,
} from 'reactstrap';

import trl from 'modules/translation';
import { EQUALITY_TYPE_IGNORE, EQUALITY_TYPE_NOT_STRICT, EQUALITY_TYPE_STRICT } from 'constants/equality';
import 'components/Header/Links/FAQ/FAQ.less';

const FAQ = () => {
    const [modal, setModal] = useState(false);
    const toggle = useCallback(() => {
        setModal(!modal);
    }, [modal]);

    return (
        <Fragment>
            <span className="faq-link" onClick={toggle}>
                {trl('Header.Links.FAQ.title')}
            </span>
            <Modal isOpen={modal} toggle={toggle} size="lg">
                <ModalHeader toggle={toggle}>FAQ</ModalHeader>
                <ModalBody>
                    <ListGroup>
                        <ListGroupItem>
                            <ListGroupItemHeading>{trl('FAQ.ForWhat.question')}</ListGroupItemHeading>
                            <ListGroupItemText>{trl('FAQ.ForWhat.answer')}</ListGroupItemText>
                        </ListGroupItem>
                        <ListGroupItem>
                            <ListGroupItemHeading>{trl('FAQ.ComparisonTypes.question')}</ListGroupItemHeading>
                            <div>
                                {trl('FAQ.ComparisonTypes.Answer.text')}
                                <ol>
                                    <li>
                                        {EQUALITY_TYPE_STRICT} - {trl('FAQ.ComparisonTypes.Answer.noStrict')}
                                    </li>
                                    <li>
                                        {EQUALITY_TYPE_NOT_STRICT} - {trl('FAQ.ComparisonTypes.Answer.strict')}
                                    </li>
                                    <li>
                                        {EQUALITY_TYPE_IGNORE} - {trl('FAQ.ComparisonTypes.Answer.ignore')}
                                    </li>
                                </ol>
                            </div>
                        </ListGroupItem>
                        <ListGroupItem>
                            <ListGroupItemHeading>{trl('FAQ.ComparisonResult.question')}</ListGroupItemHeading>
                            <div>
                                <ol>
                                    <li>
                                        <Badge color="req" pill>
                                            !
                                        </Badge>
                                        {trl('FAQ.ComparisonResult.required')}
                                    </li>
                                    <li>
                                        <Badge color="diff" pill>
                                            !
                                        </Badge>
                                        {trl('FAQ.ComparisonResult.difference')}
                                    </li>
                                </ol>
                            </div>
                        </ListGroupItem>
                    </ListGroup>
                </ModalBody>
                <ModalFooter>
                    <Button color="primary" onClick={toggle}>
                        {trl('Header.Links.FAQ.button')}
                    </Button>
                </ModalFooter>
            </Modal>
        </Fragment>
    );
};

export default FAQ;
