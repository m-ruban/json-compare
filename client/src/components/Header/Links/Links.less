@import "~styles/variables.less";

.feedback {
  color: @secondary-color !important;
}
