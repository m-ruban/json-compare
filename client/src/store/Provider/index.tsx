import * as React from 'react';

import { COMPARE_REQ } from 'constants/compare';
import { EQUALITY_TYPE_STRICT } from 'constants/equality';
import { AUTHOR_LINK, FEEDBACK_LINK } from 'constants/links';
import { LOG_INFO } from 'constants/log';
import { Provider } from 'store/Context';

import { equalAlerts } from 'modules/compare';
import { parseAndLog } from 'modules/parse';

interface IProviderProps {
    children: any;
}

interface IProviderState {
    alerts: Map<string, object>;
    author: string;
    changeEqualityTypes: (event: React.SyntheticEvent) => void;
    changeObj: (event: React.SyntheticEvent) => void;
    changeTool: (value: string) => void;
    equalityTypes: string;
    feedback: string;
    isCompare: boolean;
    log: object;
    objObject: object;
    objStr: string;
    toggleCompare: (event: React.SyntheticEvent) => void;
    toolObject: object;
    toolStr: string;
}

class AppProvider extends React.Component<IProviderProps, IProviderState> {
    constructor(props) {
        super(props);
        this.state = {
            alerts: new Map(),
            author: AUTHOR_LINK,
            changeEqualityTypes: this.changeEqualityTypes.bind(this),
            changeObj: this.changeObj.bind(this),
            changeTool: this.changeTool.bind(this),
            equalityTypes: EQUALITY_TYPE_STRICT,
            feedback: FEEDBACK_LINK,
            isCompare: false,
            log: { text: '', type: null },
            objObject: {},
            objStr: '',
            toggleCompare: this.toggleCompare.bind(this),
            toolObject: {},
            toolStr: '',
        };
    }

    public toggleCompare() {
        const log = {
            text: 'toggle compare',
            type: LOG_INFO,
        };

        this.setState({ isCompare: !this.state.isCompare, log }, this.changeAlerts);
    }

    public changeEqualityTypes(event) {
        const { value } = event.target;
        const log = {
            text: 'change equality types',
            type: LOG_INFO,
        };
        this.setState({ equalityTypes: value, log }, this.changeAlerts);
    }

    public changeTool(value: string) {
        const { obj, log } = parseAndLog(value);
        this.setState({ toolObject: obj, toolStr: value, log }, this.changeAlerts);
    }

    public changeObj(value: string) {
        const { obj, log } = parseAndLog(value);
        this.setState({ objObject: obj, objStr: value, log }, this.changeAlerts);
    }

    public changeAlerts() {
        const { toolObject, objObject, equalityTypes, isCompare } = this.state;
        const isEmpty = (obj) => Object.keys(obj).length === 0;

        if (!isEmpty(toolObject) && !isEmpty(objObject) && isCompare) {
            const alerts = equalAlerts(toolObject, objObject, equalityTypes);
            let req = 0;
            let diff = 0;
            let text = '';

            alerts.forEach((alert) => {
                if (!alert.scalar) { return; }
                if (alert.res === COMPARE_REQ) {
                    req++;
                } else {
                    diff++;
                }
            });

            if (diff > 0 || req > 0) {
                text = `${diff} alert${diff > 1 ? 's' : ''}, ${req} required`;
            } else {
                text = 'objects are equal';
            }

            this.setState({
                alerts,
                log: { text, type: LOG_INFO },
            });
        }
    }

    public render() {
        return <Provider value={this.state}>{this.props.children}</Provider>;
    }
}

export default AppProvider;
