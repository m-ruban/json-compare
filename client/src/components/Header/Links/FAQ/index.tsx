import React from 'react';
import {
    Button,
    ListGroup,
    ListGroupItem,
    ListGroupItemHeading,
    Modal,
    ModalBody,
    ModalFooter,
    ModalHeader,
} from 'reactstrap';

import 'components/Header/Links/FAQ/FAQ.less';
import Result from 'components/Header/Links/FAQ/Result';
import Types from 'components/Header/Links/FAQ/Types';
import trl from 'modules/translation';

const FAQ = () => {
    const [modal, setModal] = React.useState(false);
    const toggle = React.useCallback(() => {
        setModal(!modal);
    }, [modal]);

    return (
        <>
            <span className="faq-link" onClick={toggle}>
                {trl('Header.Links.FAQ.title')}
            </span>
            <Modal isOpen={modal} toggle={toggle} size="lg">
                <ModalHeader toggle={toggle}>FAQ</ModalHeader>
                <ModalBody>
                    <ListGroup>
                        <ListGroupItem>
                            <ListGroupItemHeading>{trl('FAQ.ForWhat.question')}</ListGroupItemHeading>
                            <div>{trl('FAQ.ForWhat.answer')}</div>
                        </ListGroupItem>
                        <ListGroupItem>
                            <ListGroupItemHeading>{trl('FAQ.ComparisonTypes.question')}</ListGroupItemHeading>
                            <Types />
                        </ListGroupItem>
                        <ListGroupItem>
                            <ListGroupItemHeading>{trl('FAQ.ComparisonResult.question')}</ListGroupItemHeading>
                            <Result />
                        </ListGroupItem>
                    </ListGroup>
                </ModalBody>
                <ModalFooter>
                    <Button color="primary" onClick={toggle}>
                        {trl('Header.Links.FAQ.button')}
                    </Button>
                </ModalFooter>
            </Modal>
        </>
    );
};

export default FAQ;
