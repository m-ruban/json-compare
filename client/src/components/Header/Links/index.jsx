import React, { useCallback, useState, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Nav, NavItem, Collapse, NavLink, NavbarToggler } from 'reactstrap';

import withConsumer from 'hoc/withConsumer';
import trl from 'modules/translation';
import Examples from 'components/Header/Links/Examples';
import FAQ from 'components/Header/Links/FAQ';
import 'components/Header/Links/Links.less';

const Links = ({ feedback }) => {
    const [isOpen, setOpen] = useState(false);
    const toggle = useCallback(() => {
        setOpen(!isOpen);
    }, [isOpen]);

    return (
        <Fragment>
            <NavbarToggler onClick={toggle} />
            <Collapse isOpen={isOpen} navbar>
                <Nav className="ml-auto" navbar>
                    <Examples />
                    <NavItem>
                        <NavLink>
                            <FAQ />
                        </NavLink>
                    </NavItem>
                    <NavItem>
                        <NavLink className="feedback" href={feedback} target="_blank">
                            {trl('Header.Links.feedback')}
                        </NavLink>
                    </NavItem>
                </Nav>
            </Collapse>
        </Fragment>
    );
};

Links.propTypes = {
    feedback: PropTypes.string,
};

export default withConsumer(Links);
