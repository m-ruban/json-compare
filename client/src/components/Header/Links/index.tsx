import React from 'react';
import { Collapse, Nav, NavbarToggler, NavItem, NavLink } from 'reactstrap';

import Examples from 'components/Header/Links/Examples';
import FAQ from 'components/Header/Links/FAQ';
import withConsumer from 'hoc/withConsumer';
import trl from 'modules/translation';

import 'components/Header/Links/Links.less';

interface ILinksProps {
    feedback: string;
}

const Links = ({ feedback }: ILinksProps) => {
    const [isOpen, setOpen] = React.useState(false);
    const toggle = React.useCallback(() => {
        setOpen(!isOpen);
    }, [isOpen]);

    return (
        <>
            <NavbarToggler onClick={toggle} />
            <Collapse isOpen={isOpen} navbar>
                <Nav className="ml-auto" navbar>
                    <Examples />
                    <NavItem>
                        <NavLink>
                            <FAQ />
                        </NavLink>
                    </NavItem>
                    <NavItem>
                        <NavLink className="feedback" href={feedback} target="_blank">
                            {trl('Header.Links.feedback')}
                        </NavLink>
                    </NavItem>
                </Nav>
            </Collapse>
        </>
    );
};

export default withConsumer(Links);
