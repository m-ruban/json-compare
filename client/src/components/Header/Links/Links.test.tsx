import React from 'react';

import { render } from 'enzyme';

import Links from 'components/Header/Links';
import { Provider } from 'store/Context';
import feedback from 'tests/mocks/feedback';

describe('<Links />', () => {
    it('Renders correctly', () => {
        const output = render(
            <Provider value={feedback}>
                <Links />
            </Provider>,
        );
        expect(output).toMatchSnapshot();
    });
});
