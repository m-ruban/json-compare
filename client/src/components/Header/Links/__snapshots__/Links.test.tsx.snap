// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Links /> Renders correctly 1`] = `
Array [
  <button
    class="navbar-toggler"
    type="button"
  >
    <span
      class="navbar-toggler-icon"
    />
  </button>,
  <div
    class="collapse navbar-collapse"
  >
    <ul
      class="ml-auto navbar-nav"
    >
      <li
        class="examples dropdown nav-item"
      >
        <a
          aria-expanded="false"
          aria-haspopup="true"
          class="examples__toggle dropdown-toggle nav-link"
          href="#"
        >
          Examples
        </a>
        <div
          aria-hidden="true"
          class="bg-dark examples__dropdown dropdown-menu"
          role="menu"
          tabindex="-1"
        >
          <button
            class="examples__dropdown-item dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Tool Example
          </button>
          <button
            class="examples__dropdown-item dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Object Example
          </button>
        </div>
      </li>
      <li
        class="nav-item"
      >
        <a
          class="nav-link"
        >
          <span
            class="faq-link"
          >
            FAQ
          </span>
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          class="feedback nav-link"
          target="_blank"
        >
          Feedback
        </a>
      </li>
    </ul>
  </div>,
]
`;
